litestar/utils/empty.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, NoReturn, TypeVar, overload
+
+from litestar.types.empty import Empty
+
+if TYPE_CHECKING:
+    from litestar.types.empty import EmptyType
+
+ValueT = TypeVar("ValueT")
+DefaultT = TypeVar("DefaultT")
+
+
+@overload
+def not_empty(value: EmptyType) -> NoReturn:
+    ...
+
+
+@overload
+def not_empty(value: EmptyType, default: EmptyType) -> NoReturn:
+    ...
+
+
+@overload
+def not_empty(value: ValueT | EmptyType) -> ValueT:
+    ...
+
+
+@overload
+def not_empty(value: ValueT | EmptyType, default: EmptyType) -> ValueT:
+    ...
+
+
+@overload
+def not_empty(value: ValueT | EmptyType, default: DefaultT) -> ValueT | DefaultT:
+    ...
+
+
+def not_empty(value: ValueT | EmptyType, default: DefaultT | EmptyType = Empty) -> ValueT | DefaultT:
```

Comment:
I find this function a bit confusing. It feels like it's doing too many things. A function named `not_empty`, I would expect to only ever return a `bool`, indicating if the value I passed in is empty.

Maybe this is a naming issue? Would something like `ensure_not_empty` make more sense?

Reply:
`ensure_not_empty` just feels wordier than it needs to be.. to me, `is_not_empty` is name of the predicate/bool returning function. This receives a value, and promises that it won't return empty.. "not empty".

That's the entirety of my argument so if you still think I should change it, I will.

Reply:
If it was just a predicate, I'd agree with you, but since it optionally returns a value it feels off.

> is name of the predicate/bool returning function

That's how I would read it, but it doesn't *always* return a bool, sometimes it returns a different value.

How about `not_empty_or`? Less wordy but it doesn't sound as much like just a predicate :eyes:

Reply:
I'm not using the raising behavior anywhere, so I'm just going to make it `value_or_default()`.